dan_layer/common_types/src/lock_intent.rs

```diff
@@ -0,0 +1,80 @@
+//   Copyright 2024 The Tari Project
+//   SPDX-License-Identifier: BSD-3-Clause
+
+use std::{fmt, str::FromStr};
+
+use tari_bor::{Deserialize, Serialize};
+use tari_engine_types::substate::SubstateId;
+
+use crate::{SubstateAddress, ToSubstateAddress, VersionedSubstateId};
+
+pub trait LockIntent {
+    fn substate_id(&self) -> &SubstateId;
+    fn lock_type(&self) -> SubstateLockType;
+    fn version_to_lock(&self) -> u32;
+    fn requested_version(&self) -> Option<u32>;
+
+    fn to_versioned_substate_id(&self) -> VersionedSubstateId {
+        VersionedSubstateId::new(self.substate_id().clone(), self.version_to_lock())
+    }
+}
+
+impl<T: LockIntent> ToSubstateAddress for T {
+    fn to_substate_address(&self) -> SubstateAddress {
+        SubstateAddress::from_substate_id(self.substate_id(), self.version_to_lock())
+    }
+}
+
+/// Substate lock flags
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize)]
+#[cfg_attr(
+    feature = "ts",
+    derive(ts_rs::TS),
+    ts(export, export_to = "../../bindings/src/types/")
+)]
```

Comment:
i notice [ts-rs](https://docs.rs/ts-rs/latest/ts_rs/trait.TS.html) is used but when do we need to export ts bindings generally?

Reply:
Basically when it is part of the public API in any way. For this type it is part of a chain of types that are in the Transaction and Block types. Since you can query for Block and Transactions using JSON-RPC we need ts_rs. Rust will not compile if TS rs is used on a "parent" type and not on a child type so you'll never need to figure that out for yourself. You can also manually define TS types on fields e.g.

```rust
    #[cfg_attr(feature = "ts", ts(type = "string"))]
    pub public_key: PublicKey,
```